packages/flutter/lib/src/cupertino/text_selection.dart

```diff
@@ -41,97 +43,149 @@ const TextStyle _kToolbarButtonFontStyle = TextStyle(
   color: CupertinoColors.white,
 );
 
+/// The direction of the triangle attached to the toolbar.
+///
+/// Defaults to showing the triangle downwards if sufficient space is available
+/// to show the toolbar above the text field. Otherwise, the toolbar will
+/// appear below the text field and the triangle's direction will be [up].
+enum _ArrowDirection { up, down }
+
 /// Paints a triangle below the toolbar.
 class _TextSelectionToolbarNotchPainter extends CustomPainter {
+  const _TextSelectionToolbarNotchPainter(
+    this.arrowDirection,
+    this.triangleOffsetX,
```

Comment:
I see 3 names here for the same thing: notch, arrow, and triangle.  It looks like we use multiple names in the existing code too though :( Maybe try to choose the most common one and stick with it if you think it will help.